### PR TITLE
#calendar522 add change participation method

### PR DESCRIPTION
**Pull request by asmaNode**

### CODE REVIEW

File: lib/api/calendars.ts [hunk 1]
```diff
@@ -51,3 +51,12 @@ export const modifyEvent = (client: DAVClient) => async (eventPath: string, even
...
   });
 }
+
+export const changeParticipation = (client: DAVClient) => async (eventPath: string , mailAttendee:string,action:string , event:CalendarEventObject): Promise<string> => {
+  if(event.attendees){
+    const  attendeeIndex = event.attendees.findIndex((attendee => attendee.email == mailAttendee));
+    event.attendees[attendeeIndex].partstat = action;
+  }
+
+  return  await modifyEvent(client)(eventPath,event);
```
Comment (rezk2ll) [inline lib/api/calendars.ts:61]:
```suggestion
  
  return  await modifyEvent(client)(eventPath,event);
  }
```

no point in calling modifyEvent if nothing changed

File: lib/api/calendars.ts [hunk 2]
```diff
@@ -51,3 +51,17 @@ export const modifyEvent = (client: DAVClient) => async (eventPath: string, even
     body: translate(event)
   });
 }
+
+export const changeParticipation = (client: DAVClient) => async (eventPath: string, mailAttendee: string, action: string, event: CalendarEventObject): Promise<string> => {
```
Comment (huy-ta) [inline lib/api/calendars.ts:55]: The arguments are long enough to be converted into an object argument.

File: lib/api/calendars.ts [hunk 3]
```diff
@@ -51,3 +51,17 @@ export const modifyEvent = (client: DAVClient) => async (eventPath: string, even
...
+    event.attendees.map(attendee => {
+      if (attendee.email == mailAttendee) {
+        attendee.partstat = action;
+       
+        return  modifyEvent(client)(eventPath, event);
+      }
+    })
+  }
+
+  return Promise.reject('can not change participation')
```
Comment (huy-ta) [inline lib/api/calendars.ts:66]:
```suggestion
  if (!Array.isArray(event.attendees)) return Promise.reject('Can not change participation');

  for (const attendee of event.attendees) {
    if (attendee.email !== attendeeEmail) continue;
    
    attendee.partstat = action;
    
    return modifyEvent(client)(eventPath, event);
  }

  return Promise.reject('No matching attendee found in the event')
```

Comment (huy-ta) [reply lib/api/calendars.ts:66]: You should avoid nesting `if`s whenever possible. The code will be much more readable without nested `if`s.

File: test/api/calendar.test.ts [hunk 4]
```diff
@@ -124,3 +124,117 @@ END:VCALENDAR',
...
+      description: 'simple description',
+      location: 'some location',
+      duration: {
+        hours: 1,
+      },
+      extendedProps: {},
+    });
+  });
+
+  it('should return error message if event does not containt attendee ', () => {
```
Comment (huy-ta) [inline test/api/calendar.test.ts:191]:
```suggestion
  it('should return error message if event does not contain any attendees', () => {
```

Comment (huy-ta) [reply test/api/calendar.test.ts:191]: Please mind the spaces.

File: test/api/calendar.test.ts [hunk 5]
```diff
@@ -124,3 +124,117 @@ END:VCALENDAR',
...
+        hours: 1,
+      },
+      extendedProps: {},
+    };
+
+    changeParticipation(davClient)(path, mailAttendee, newAction, testEventObject);
+    expect(changeParticipation(davClient)(path, mailAttendee, newAction, testEventObject)).rejects.toEqual('can not change participation')
+  });
+
+  it('should return error message if user does exist ', () => {
```
Comment (huy-ta) [inline test/api/calendar.test.ts:213]:
```suggestion
  it('should reject if the attendee does exist in the event', () => {
```

File: test/api/calendar.test.ts [hunk 6]
```diff
@@ -124,3 +124,117 @@ END:VCALENDAR',
...
+      duration: {
+        hours: 1,
+      },
+      extendedProps: {},
+    };
+
+    changeParticipation(davClient)(path, mailAttendee, newAction, testEventObject);
+    expect(modifyEvent).toHaveBeenCalled;
+    expect(testEventObject).toStrictEqual({
+
```
Comment (huy-ta) [inline test/api/calendar.test.ts:171]:
```suggestion
```

File: lib/api/calendars.ts (hunk 2 above)
Comment (asmaNode) [reply lib/api/calendars.ts]: all the arguments is required to be passed in this case ! what we should do

Comment (huy-ta) [reply lib/api/calendars.ts]: Convert it to an object argument.

```
export interface ChangeParticipationOptions {
  eventPath: string;
  attendeeEmail: string;
  action: string;
  event: CalendarEventObject;
}

export const changeParticipation = (client: DAVClient) => async ({ eventPath, mailAttendee, action, event }: ChangeParticipationOptions): Promise<string> => {
```

You can read https://github.com/ryanmcdermott/clean-code-javascript#function-arguments-2-or-fewer-ideally for more information about this rule.

File: test/api/calendar.test.ts [hunk 7]
```diff
@@ -143,6 +143,102 @@ describe('the deleteEvent method', () => {
...
+        hours: 1,
+      },
+      extendedProps: {},
+    };
+
+    changeParticipation(davClient)(path, mailAttendee, newAction, testEventObject);
+    expect(changeParticipation(davClient)(path, mailAttendee, newAction, testEventObject)).rejects.toEqual('Can not change participation')
+  });
+
+  it('should reject if the attendee does exist in the event', () => {
```
Comment (huy-ta) [inline test/api/calendar.test.ts:213]:
```suggestion
  it('should reject if the attendee does not exist in the event', () => {
```

**Comment by huy-ta** [comment]
There should only be one commit.

File: test/api/calendar.test.ts [hunk 8]
```diff
@@ -143,6 +143,116 @@ describe('the deleteEvent method', () => {
...
+      },
+      extendedProps: {},
+    };
+
+    const testChangeParticipationObject: ChangeParticipationOptions = {
+      eventPath: 'events/123456.ics',
+      attendeeEmail: 'anotherUser@gmail.com',
+      action: 'accept',
+      event: testEventObject
+
```
Comment (rezk2ll) [inline test/api/calendar.test.ts:247]:
```suggestion
```

File: test/api/calendar.test.ts [hunk 9]
```diff
@@ -143,6 +143,116 @@ describe('the deleteEvent method', () => {
...
+      },
+      extendedProps: {},
+    };
+
+    const testChangeParticipationObject: ChangeParticipationOptions = {
+      eventPath: 'events/123456.ics',
+      attendeeEmail: 'test1@gmail.com',
+      action: 'accept',
+      event: testEventObject
+
```
Comment (rezk2ll) [inline test/api/calendar.test.ts:215]:
```suggestion
```

**Review by rezk2ll — APPROVED** [review]

**Comment by huy-ta** [comment]
There should only be 1 commit.

**Review by huy-ta — APPROVED** [review]

**Review by renaudboyer — APPROVED** [review]

**Pull request merged**